proposals/0000-existentials.rst

```diff
@@ -0,0 +1,504 @@
+   is the same as the representation of ``ty``. This is exactly the same as the treatment
+   of ``forall`` and type application.
+
+Optional Extension
+------------------
+
+Extend the treatment of existentials to work in types. (Do *not* extend the treatment
+of constraints with ``/\``.) This would affect the kind
+inference algorithm analogously to the effects on the type inference algorithm above.
+The only complication is that we would need ``TWitness`` instead of ``Witness``.
+A ``TWitness`` accepts a type, not an expression. (In the glorious future of Dependent
+Haskell, ``TWitness`` becomes a synonym for ``Witness``.)
+
+Supporting this feature would require yet more extensions to Core:
+* ``PackTy`` and ``OpenTy`` in ``Type``, analogous to ``Pack`` and ``Open`` in ``Expr``.
+* ``TWitness`` in ``Type``, analogous to ``Witness``.
+* ``TWitnessPackCo`` and ``TWitnessCo`` in ``Coercion``, analogous to ``WitnessPackCo`` and ``WitnessCo``.
+
+I would prefer not to have this extension, but save type-level support for packing
+and opening existentials until we have proper dependent types.
+
+Examples
+--------
+
+1. See the filter_ example, above.
+
+#. ::
+
+     toVec :: [a] -> exists n. KnownNat n /\ Vec n a
+     toVec []     = VNil
+     toVec (x:xs) = x :> toVec xs
+
+     -- NB: requires inputs to have the same length
+     zipWithV :: (a -> b -> c) -> Vec n a -> Vec n b -> Vec n c
+     zipWithV = ...
+
+     -- from GHC.TypeLits:
+     sameNat :: forall a b proxy1 proxy2. (KnownNat a, KnownNat b) => proxy1 a -> proxy2 b -> Maybe (a :~: b)
+
+     -- returns first argument iff its length matches that of the second, in O(1) time
+     sameLengthAs :: forall m n a b. (KnownNat m, KnownNat n) => Vec m a -> Vec n b -> Maybe (Vec n a)
+     sameLengthAs v1 _ = do Refl <- sameNat @m @n Proxy Proxy
+                            return v1
+
+     frob :: (a -> b -> c) -> [a] -> [b] -> exists n. Maybe (Vec n c)
+     frob f as bs = do let as' = toVec as
+                           bs' = toVec bs
+                       as'' <- sameLengthAs as' bs'
+                       return (zipWithV f as'' bs')
+
+#. Imagine this within GHC::
+
+     -- in GHC.Utils.Outputable
+     sep :: [SDoc] -> SDoc
+     ppr :: Outputable a => a -> SDoc
+
+     pprs :: [exists a. Outputable a /\ a] -> SDoc
+     pprs = sep . map ppr    -- this type-checks because of the rules for class constraints on existentials
+
+     someTypecheckingFunction a b c d e f = do ...
+                                               traceTc "herald" (pprs [a, b, c, d, e, f])
+                                               -- today, we'd need [ppr a, ppr b, ppr c, ppr d, ppr e, ppr f],
+                                               -- which is cluttersome.
+                                               ...
+
+Effect and Interactions
+-----------------------
+
+1. Programming with existentials is now straightforward and uncluttered. This
```

Comment:
You are underselling it. “straightforward and uncluttered” is an improvement, but only in style. The proposed feature has more important operational implications:

* Existential quantification can be erased. Currently it requires boxing, therefore a pointer indirection.
* Laziness can be preserved. Currently, operating on existentially quantified values requires forcing them.

Not only that, this proposal would help with GHC ticket [#1965](https://gitlab.haskell.org/ghc/ghc/-/issues/1965) by adding the required machinery to Core. Then we could optimize existing definitions in `base`, e.g. `SomeTypeRep` could become a newtype. This would be a performance win even for users who are unaware of the new feature.

Reply:
Yes. Good point. Thanks!